// melphis/main.js
function Clave() {
    let clave = Math.random().toString(36).slice(6).split('').map((x, i) => x[(i % 2) ? 'toLowerCase' : 'toUpperCase']()).join('')
    return clave
}

function Kasake() {
    let goat = ['https://www.lavanguardia.com/r/GODO/LV/p6/WebSite/2019/06/26/Recortada/img_npalou_20180829-191719_imagenes_lv_otras_fuentes_goatspreferh-ktSB-U463127665413HSB-992x558@LaVanguardia-Web.jpg', 'https://www.petalatino.com/wp-content/uploads/Goats-maximili.jpg', 'https://d3dv6ncga0u0ql.cloudfront.net/wp-content/uploads/2017/06/Selecting-Goats-for-Milk-Meat.jpg', 'https://static.nationalgeographicla.com/files/styles/image_3200/public/01-goat-emotions-nationalgeographic_2165409.jpg?w=1600&h=900', 'https://cdn.bioguia.com/embed/5393a9454058114aa05761fb447550ae9ad4aec31570463061/video-cabras-arana.jpg']
    let randomGoat = goat[Math.floor(Math.random() * goat.length)]  
    return randomGoat
}

class KasakeWapo {
    constructor(Nivel, Centimetros) {
        this.Nivel = Nivel
        this.Pene = Centimetros
    }

    Estadisticas() {
        `Kasake tiene un Pene de ${this.Pene}, Su Nivel esta por ${this.Nivel}`
    }
}

const Stats = new KasakeWapo('Por Encima de los Dioses', '60cm 😲')


module.exports = {

clave: Clave,
kasake: Kasake

}
